### PR TITLE
Make SANITIZER_MIN_OSX_VERSION a cache variable

### PR DESCRIPTION
**Pull request by cjappl**
It is desirable to be able to configure the `-mmacosx-version-min` flag for the sanitizers, but this flag was never made a CACHE variable in cmake.

By doing this, it will allow developers to select different minimum versions, which results in different interceptors being enabled or disabled on their platforms. This version can now persist between cmake runs, so it can be remembered by cmake, and edited in the cache file.

### CODE REVIEW

**Comment by cjappl** [comment]
@vitalybuka @ramosian-glider @kcc 

Kindly requesting a review on this patch. Thanks in advance

**Comment by cjappl** [comment]
@vitalybuka @ramosian-glider  @kcc 

Ping

**Comment by cjappl** [comment]
@vitalybuka @ramosian-glider @kcc @petrhosek 

Ping

**Review by vitalybuka — APPROVED** [review]
LGTM but please wait a few days for others feedback

@petrhosek For CMAKE expertise 

@yln @sundahl as it impacts OSX

File: compiler-rt/cmake/config-ix.cmake
Comment (petrhosek) [inline compiler-rt/cmake/config-ix.cmake:461]: Could we reverse the order, that is compute the default value in the block below, and then set the cached value as
```cmake
set(SANITIZER_MIN_OSX_VERSION "${MIN_OSX_VERSION}" CACHE STRING
    "Minimum OS X version to target (e.g. 10.10) for sanitizers.")
```
after the block?

Comment (cjappl) [reply compiler-rt/cmake/config-ix.cmake]: Addressed!

File: compiler-rt/cmake/config-ix.cmake
Comment (petrhosek) [inline compiler-rt/cmake/config-ix.cmake:488]: Can you move out of the `if` block (past the `endif`)?

Comment (cjappl) [reply compiler-rt/cmake/config-ix.cmake]: Yep, done! Thanks for the clarification, I didn't grok it the first time.

**Review by petrhosek — APPROVED** [review]

**Review by yln — APPROVED** [review]
Let me summarize my understanding: the intention here is to promote `SANITIZER_MIN_OSX_VERSION` to a CMAKE cache variable.  We don't change any behavior or default.

LGTM, thanks!  

@rsundahl @thetruestblue
At some point we should think about bumping the minimal sanitizer target enabling some cleanups.

**Comment by cjappl** [comment]
> Let me summarize my understanding: the intention here is to promote SANITIZER_MIN_OSX_VERSION to a CMAKE cache variable. We don't change any behavior or default.

Your understanding is correct! No changes in behavior, just allowing people to tweak it easier.

> At some point we should think about bumping the minimal sanitizer target enabling some cleanups.

One of the main reasons behind me putting up this review is I found a handful of things that would be cleaned up (allowing sanitizers to hook into calls introduced in newer MacOS versions). Once this is merged, I'm going to put up a few more in this series based on what I found! I'll make sure to add you to those reviews to get your thoughts @yln . 

Thanks for the review.

**Comment by cjappl** [comment]
Pinging either:

@rsundahl for a review

Or one of my previous reviewers with write access for a merge, if appropriate.

@petrhosek @yln @vitalybuka 

Thanks!

**Comment by cjappl** [comment]
Pinging either:

@rsundahl for a review

Or one of my previous reviewers with write access for a merge, if appropriate.

@petrhosek @yln @vitalybuka

**Comment by jfgoog** [comment]
This does actually change behavior. If `-mmacosx-version-min` is set, cmake will behave differently between the first and subsequent runs. More info at https://github.com/rust-lang/rust/pull/122504#issuecomment-2035091552

**Comment by cjappl** [comment]
Agreed on your analysis @jfgoog . I see that on subsequent runs, `MACOSX_VERSION_MIN_FLAG` will be unset. This was an unintended side effect (bug).

I think this patch would fix the problem, while maintaining the flexibility of this PR? Essentially hoisting the regex match out of the if statement. WDYT?

```
diff --git a/compiler-rt/cmake/config-ix.cmake b/compiler-rt/cmake/config-ix.cmake
index d5dc4a40c097..b2cae81d5032 100644
--- a/compiler-rt/cmake/config-ix.cmake
+++ b/compiler-rt/cmake/config-ix.cmake
@@ -463,9 +463,10 @@ if(APPLE)

   set(DEFAULT_SANITIZER_MIN_OSX_VERSION 10.10)
   set(DARWIN_osx_MIN_VER_FLAG "-mmacosx-version-min")
+  string(REGEX MATCH "${DARWIN_osx_MIN_VER_FLAG}=([.0-9]+)"
+         MACOSX_VERSION_MIN_FLAG "${CMAKE_CXX_FLAGS}")
+
   if(SANITIZER_MIN_OSX_VERSION STREQUAL "")
-    string(REGEX MATCH "${DARWIN_osx_MIN_VER_FLAG}=([.0-9]+)"
-           MACOSX_VERSION_MIN_FLAG "${CMAKE_CXX_FLAGS}")
     if(MACOSX_VERSION_MIN_FLAG)
       set(MIN_OSX_VERSION "${CMAKE_MATCH_1}")
     elseif(CMAKE_OSX_DEPLOYMENT_TARGET)
```

**Comment by jfgoog** [comment]
That seems reasonable, but I am hardly an expert here.

The underlying issue for me, though, which pre-dates your change, is that it seems like if the `-mmacosx-version-min` flag is set, we will never detect any architectures for compiler-rt, essentially making it impossible to build anything. Is that intended behavior?

**Comment by cjappl** [comment]
Pinging a couple experts to weigh in (as I am hardly an expert either)

@yln @vitalybuka @petrhosek 

Pending on when they get back to us, if this is a blocking issue for rust you should file a bug report for this. I wouldn't want this "buried" in this PR with just us two talking about it

**Comment by jfgoog** [comment]
It's not blocking. It only happens with newer versions of the cc crate, and the rustc bootstrap code has an old version pinned that doesn't have this problem.

**Comment by cjappl** [comment]
I put up a review to at least revert this to the previous way of doing things:

https://github.com/llvm/llvm-project/pull/87580

I understand this doesn't fix your "bigger" issue, but I agree that something about that smells fishy.

**Pull request merged**
